src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonObjectDefaultConverter.cs

```diff
@@ -0,0 +1,415 @@
+
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    /// <summary>
+    /// Default base class implementation of <cref>JsonObjectConverter{T}</cref>.
+    /// </summary>
+    internal sealed class JsonObjectDefaultConverter<T> : JsonObjectConverter<T>
+    {
+        internal override bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options, ref ReadStack state, out T value)
+        {
+            bool shouldReadPreservedReferences = options.ReferenceHandling.ShouldReadPreservedReferences();
+            object obj;
+
+            if (!state.SupportContinuation && !shouldReadPreservedReferences)
+            {
+                // Fast path that avoids maintaining state variables and dealing with preserved references.
+
+                if (reader.TokenType != JsonTokenType.StartObject)
+                {
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                }
+
+                if (state.Current.JsonClassInfo.CreateObject == null)
+                {
+                    ThrowHelper.ThrowNotSupportedException_DeserializeNoParameterlessConstructor(state.Current.JsonClassInfo.Type);
+                }
+
+                obj = state.Current.JsonClassInfo.CreateObject!()!;
+
+                // Read all properties.
+                while (true)
+                {
+                    // Read the property name or EndObject.
+                    reader.Read();
+
+                    JsonTokenType tokenType = reader.TokenType;
+                    if (tokenType == JsonTokenType.EndObject)
+                    {
+                        break;
+                    }
+
+                    if (tokenType != JsonTokenType.PropertyName)
+                    {
+                        ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                    }
+
+                    JsonSerializer.LookupProperty(
+                        obj,
+                        ref reader,
+                        options,
+                        ref state,
+                        out JsonPropertyInfo jsonPropertyInfo,
+                        out bool useExtensionProperty);
+
+                    // Skip the property if not found.
+                    if (!jsonPropertyInfo.ShouldDeserialize)
+                    {
+                        reader.Skip();
+                        continue;
+                    }
+
+                    // Set the property value.
+                    reader.Read();
+
+                    if (!useExtensionProperty)
+                    {
+                        jsonPropertyInfo.ReadJsonAndSetMember(obj, ref state, ref reader);
+                    }
+                    else
+                    {
+                        jsonPropertyInfo.ReadJsonAndAddExtensionProperty(obj, ref state, ref reader);
+                    }
```

Comment:
You are missing a call to `state.Current.EndProperty`, similar to the slow path on line 252.
Doing that fixes https://github.com/dotnet/runtime/pull/2259/files#r375585778.